File: playwright-csharp/Pages/AuthenticationPage.cs
using Fictoshop.PlaywrightTests.Support;
using Microsoft.Playwright;
using NUnit.Framework;
using static Microsoft.Playwright.Assertions;

namespace Fictoshop.PlaywrightTests.Pages;

public sealed class AuthenticationPage
{
    private readonly IPage _page;
    private readonly ILocator _username;
    private readonly ILocator _password;
    private readonly ILocator _loginButton;

    public AuthenticationPage(IPage page)
    {
        _page = page;
        _username = page.GetByLabel("Username");
        _password = page.GetByLabel("Password");
        _loginButton = page.GetByRole(AriaRole.Button, new() { Name = "Log in" });
    }

    public async Task OpenFromStorefrontAsync()
    {
        await _page.GotoAsync(TestSettings.BaseUrl);
        await _page.Locator("header").GetByRole(AriaRole.Link, new() { Name = "Log in", Exact = true }).ClickAsync();
    }

    public Task OpenAsync() => _page.GotoAsync(TestSettings.BaseUrl + "/signin");
    public Task SubmitAsync() => _loginButton.ClickAsync();

    public async Task SubmitInvalidAsync()
    {
        await _username.FillAsync("unknown-playwright-user");
        await _password.FillAsync("Invalid-Playwright-Password!");
        await SubmitAsync();
    }

    public async Task LoginAsync(Credentials credentials)
    {
        await OpenAsync();
        await _username.FillAsync(credentials.Username);
        await _password.FillAsync(credentials.Password);
        await SubmitAsync();
        var destination = credentials.Username.Contains("_admin_", StringComparison.Ordinal)
            ? TestSettings.BaseUrl + "/admin/"
            : TestSettings.BaseUrl + "/";
        await _page.WaitForURLAsync(destination);
    }

    public async Task AssertCredentialFieldsAsync()
    {
        await Expect(_page).ToHaveURLAsync(TestSettings.BaseUrl + "/signin");
        await Expect(_username).ToBeVisibleAsync();
        await Expect(_password).ToBeVisibleAsync();
    }

    public Task AssertLoginButtonAsync() => Expect(_loginButton).ToBeVisibleAsync();

    public async Task AssertMissingCredentialsAsync()
    {
        Assert.That(await _username.EvaluateAsync<bool>("element => element.validity.valueMissing"), Is.True);
        Assert.That(await _username.EvaluateAsync<string>("element => element.validationMessage"), Is.Not.Empty);
    }

    public async Task AssertSignedOutAsync()
    {
        await _page.GotoAsync(TestSettings.BaseUrl);
        await Expect(_page.Locator("header").GetByRole(AriaRole.Link, new() { Name = "Log in", Exact = true })).ToBeVisibleAsync();
    }

    public Task AssertInvalidCredentialsAsync() => Expect(_page.GetByRole(AriaRole.Status)).ToHaveTextAsync("Invalid credentials");

    public async Task AssertSignInPageAsync()
    {
        await Expect(_page).ToHaveURLAsync(TestSettings.BaseUrl + "/signin");
        await Expect(_username).ToBeVisibleAsync();
        await Expect(_password).ToBeVisibleAsync();
    }
}

File: playwright-typescript/src/pages/authentication-page.ts
import assert from "node:assert/strict";
import { expect, type Locator, type Page } from "@playwright/test";
import type { Credentials } from "../support/models.js";
import { settings } from "../support/settings.js";

export class AuthenticationPage {
  private readonly username: Locator;
  private readonly password: Locator;
  private readonly loginButton: Locator;

  constructor(private readonly page: Page) {
    this.username = page.getByLabel("Username");
    this.password = page.getByLabel("Password");
    this.loginButton = page.getByRole("button", { name: "Log in" });
  }

  async openFromStorefront(): Promise<void> {
    await this.page.goto(settings.baseUrl);
    await this.page.locator("header").getByRole("link", { name: "Log in", exact: true }).click();
  }

  open(): Promise<unknown> { return this.page.goto(`${settings.baseUrl}/signin`); }
  submit(): Promise<void> { return this.loginButton.click(); }

  async submitInvalid(): Promise<void> {
    await this.username.fill("unknown-playwright-user");
    await this.password.fill("Invalid-Playwright-Password!");
    await this.submit();
  }

  async login(credentials: Credentials): Promise<void> {
    await this.open();
    await this.username.fill(credentials.username);
    await this.password.fill(credentials.password);
    await this.submit();
    const destination = credentials.superuser ? `${settings.baseUrl}/admin/` : `${settings.baseUrl}/`;
    await this.page.waitForURL(destination);
  }

  async expectCredentialFields(): Promise<void> {
    await expect(this.page).toHaveURL(`${settings.baseUrl}/signin`);
    await expect(this.username).toBeVisible();
    await expect(this.password).toBeVisible();
  }

  expectLoginButton(): Promise<void> { return expect(this.loginButton).toBeVisible(); }

  async expectMissingCredentials(): Promise<void> {
    assert.equal(await this.username.evaluate((element: HTMLInputElement) => element.validity.valueMissing), true);
    assert.ok(await this.username.evaluate((element: HTMLInputElement) => element.validationMessage));
  }

  async expectSignedOut(): Promise<void> {
    await this.page.goto(settings.baseUrl);
    await expect(this.page.locator("header").getByRole("link", { name: "Log in", exact: true })).toBeVisible();
  }

  expectInvalidCredentials(): Promise<void> { return expect(this.page.getByRole("status")).toHaveText("Invalid credentials"); }

  async expectSignInPage(): Promise<void> {
    await expect(this.page).toHaveURL(`${settings.baseUrl}/signin`);
    await expect(this.username).toBeVisible();
    await expect(this.password).toBeVisible();
  }
}
